include/boost/unordered/detail/foa/core.hpp

```diff
@@ -920,8 +920,8 @@ inline unsigned int unchecked_countr_zero(int x)
  * allocators.
  */
 
-template<typename Group,std::size_t Size>
-Group* dummy_groups()
+template<typename GroupPtr,std::size_t Size>
+GroupPtr dummy_groups()
```

Comment:
I remember why I did this funny now. It's because of this:
```cpp
if(!n&&std::is_same<group_type*,group_type_pointer>::value){
  arrays.groups_=dummy_groups<group_type_pointer,size_policy::min_size()>();
}
```

If `dummy_groups` returns a `Group*`, there's issue when `arrays.groups_` is fancy. It requires an implicit conversion on the fancy pointer side which I'm not sure we wanna do.

I'm not 100% sure how you wanna handle this. We could keep `dummy_groups` as-is and have it return a raw pointer and move move the `pointer_to` logic into the assignment at `arrays.groups_=...`.

Reply:
Yes, the code should be refactored to something like:
```cpp
static group_type_pointer new_groups(allocator_type al,std::size_t n)
{
  return new_groups(
    al,n,
    std::integral_constant<
      bool,std::is_same<group_type*,group_type_pointer>::value>{});
}

static group_type_pointer new_groups(
  allocator_type al,std::size_t n, std::false_type /* always allocate */)
{
  auto sal=allocator_type(al);
  arrays.elements_=storage_traits::allocate(sal,buffer_size(groups_size));
  ...
}

static group_type_pointer new_groups(
  allocator_type al,std::size_t n, std::true_type /* optimize for n==0 */)
{
  if(!n){
    return dummy_groups<group_type,size_policy::min_size()>();
  }
  else{
    return new_groups(al,n,std::false_type{});
  }
}
```